### PR TITLE
feat!: Full async networking

### PR DESCRIPTION
**Pull request by cmroche**
Replaced the old send then receive approach to networking with asyncio.DatagramProtocol based classes

This removed a lot of duplicate network code as discovery was already using this approach, and allows handling on device responses when they are received as opposed to waiting on them explicitly.

BREAKING CHANGE:
While interfaces are unchanged, you can no longer expect a device response is already received when `push_state_update` or `bind` have returned.

Responses will come through their appropriate handlers once they have been received.

### CODE REVIEW

**Comment by engrbm87** [comment]
Hell @cmroche , I have been working on making this library fully type hinted. I was about to make a PR for your review then I saw these changes. Can you let me know if this PR is ready to be merged so I can rebase and update my work.
Also we need to add support for new Gree Wifi modules running firmware 1.2.1. These use GCM encoding and I think we can incorporate this into the library in the encrypt/decrypt methods (also a new key is used). So my plan was to work on the same only after finishing the type hints.

File: greeclimate/network.py [hunk 1]
```diff
@@ -95,16 +117,39 @@ def resume_writing(self) -> None:
...
     def datagram_received(self, data: bytes, addr: IPAddr) -> None:
         """Handle an incoming datagram."""
         if len(data) == 0:
             return
 
         obj = json.loads(data)
-        key = GENERIC_KEY if obj.get("i") == 1 else self._key
+
+        # It could be either a v1 or v2 key
+        key = GENERIC_KEY[0] if obj.get("i") == 1 else self._key
```
Comment (engrbm87) [inline greeclimate/network.py:149]: Sorry to interrupt your coding flow, but I can see that you incorporating the new KEY and encoding for new models with firmware 1.2.1. Would you mind if I through in a few suggestions. I have a new Gree AC and I would like to be able to control it using HA.

Comment (cmroche) [reply greeclimate/network.py:148]: The plan is to add the GCM encryption support following this change, it's not actually implemented yet though.

To be pedantic. V1.2.1 sounds like you're referring to the hardware revision, my device reports 1.2.1 and uses the older encryption technique. We won't be able to select based on that version reported in the scan result.

**Comment by cmroche** [comment]
> Hell @cmroche , I have been working on making this library fully type hinted. I was about to make a PR for your review then I saw these changes. Can you let me know if this PR is ready to be merged so I can rebase and update my work. Also we need to add support for new Gree Wifi modules running firmware 1.2.1. These use GCM encoding and I think we can incorporate this into the library in the encrypt/decrypt methods (also a new key is used). So my plan was to work on the same only after finishing the type hints.

Awesome. Been working slowly to add typehints, so will be greatly appreciated.

File: greeclimate/network.py (hunk 1 above)
Comment (engrbm87) [reply greeclimate/network.py:148]: V1.2.1 is the firmware version showing on Gree app. 
The best approach is to try the GCM if the current encryption causes a timeout. 
If both make a timeout then we raise a timeout.

Comment (cmroche) [reply greeclimate/network.py:148]: Interesting... they have a weird concept of versions, but yes. Timeout approach is what I was thinking too.

That said, could you share responses from the scan and bind commands from your 1.2.1 devices please? Maybe there is something in there, I can compare with examples I have on hand.

**Comment by cmroche** [comment]
> Hell @cmroche , I have been working on making this library fully type hinted. I was about to make a PR for your review then I saw these changes. Can you let me know if this PR is ready to be merged so I can rebase and update my work. Also we need to add support for new Gree Wifi modules running firmware 1.2.1. These use GCM encoding and I think we can incorporate this into the library in the encrypt/decrypt methods (also a new key is used). So my plan was to work on the same only after finishing the type hints.

I am done with the big rework. Next I will look at the GCM encryption this week, perhaps I could get you to test it for me once it's ready as I do not have one of these devices.

File: greeclimate/network.py (hunk 1 above)
Comment (engrbm87) [reply greeclimate/network.py:148]: Here is the response from the scan 
```json
{"t": "pack", "i": 1, "uid": 0, "cid": "", "tcid": "", "pack": {"t": "dev", "cid": "9424b8bb14bc", "bc": "00000000000000000000000000000000", "brand": "gree", "catalog": "gree", "mac": "9424b8bb14bc", "mid": "10001", "model": "gree", "name": "", "lock": 0, "series": "gree", "vender": "1", "ver": "V2.0.0", "ModelType": "32772", "hid": "362001065279+U-WB05RT13V1.21.bin"}}
```
hid is showing `V1.21`

**Pull request merged**